### PR TITLE
Move pragma detection/insertion to plugins

### LINKED ISSUE #3557: --insert-pragma breaks GraphQL, CSS
**Prettier 1.9.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBHArnATgTwAQCKmuewAOlHngJYAmFAvhSADQgQAOM10AzsqACGWLBADuABWEJ+KQQDcIdViEG8YyAGaCANrzhsARlkFgA1nBgBlDqepQA5shhZMbWhDBbd+tvf1YMBImDgC2gs6uBiAAVrwAHgBCJuaWVoKhcAAy9nDeetEQ6DAcxQBM+b4gtlgByCAOJhwAFqg6KhxY9jAA6nQwzcgAHAAMbJ0Q+j1N9Z1wAfJ5bFhwGNQrwYJhEUjaBWz6odSRbiC89g46cEQQ8JXRMIKGfbQDyGVsLoLUOhcAwhBQuF6lBoEsQOh9AAVJ6yPb6BgMIA)
```sh
--insert-pragma
--parser graphql
```

**Input:**
```graphql
query Query {
  id
}

```

**Output:**
```graphql
SyntaxError: Syntax Error GraphQL request (1:1) Cannot parse the unexpected character "/".

1: /** @format */
   ^
2: 
 (1:1)
> 1 | /** @format */
    | ^
  2 | 
  3 | query Query {
  4 |   id
```

---
**Prettier 1.9.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBiABACzgGxxEAGhAgAcYBLaAZ2VAEMAnRiAdwAUmFaV6A3CBQAmREPWoxkAM3o5qcYgCNG9MAGs4MAMqlVFKAHNkMRgFcFIIRDDTZ84vvmMY7FQYC29Y2YsAragAeAEIq6ppa9O5wADL6cLZyFhCmMKQpAEwJ9iC6jE7IIJ6MalasUKKkjPowAOrCMJjIABwADMSVEPI1KqQFlXBOfPHEjHAAjqYUo670Hl5IMonE8u4U3ubL+gY4cACKphDwWRYw9Ip1Qg3I6cQm9BQ4WwDCEO6eBVDQwyCm8gAqZx4i3kAF8QUA)
```sh
--insert-pragma
--parser markdown
```

**Input:**
```markdown
# hello
```

**Output:**
```markdown
/\*_ @format _/

# hello

```

---
**Prettier 1.9.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEwAEAdEAzCEtKYgBGAhgE5ZoC+IANCBAA4wCW0AzsqBeRAO4AFCgi4pSANwisAJvRCkOMZNlIAbDnAbFypMAGs4MAMpM9rKAHNkMcgFctIGRDAr1mhhc3kYg3ZYBbUht7RwArDgAPACFdAyNjUgC4ABkLODcNRwg7GCZcgCZMjxAzcm9kEAjoeSZyCxgAdVkYAAtkAA4ABgY6iE1G3SZKurhvCQyGcjgARztWab9SQOCkVSyGTQDWEIdNi0s1OABFOwh4YscYUmJmmTbkAoZbUlY1A4BhCACgyqhoSYgOyaAAqNzE6001GoQA)
```sh
--insert-pragma
--parser json
```

**Input:**
```json
{ "foo": "bar" }
```

**Output:**
```json
/** @format */

{ "foo": "bar" }

```

### PR DESCRIPTION
**Pull request by duailibe**
Closes #3557

### CODE REVIEW

**Comment by duailibe** [comment]
This is still WIP but I'd like help/feedback on how to detect/insert pragmas in GraphQL/CSS (guess we could still use `jest-docblock`)/Markdown (also my RegEx-fu is not that great).

e.g. Should we 

```graphql
# @format

# or should we accept:

# Some comments
# in the top
#
# @something-else (are those even used in GraphQL?)
# @format
```

**Comment by azz** [comment]
Nice stuff!

Not sure if `hasPragma` should be on the `parser` or `printer` object. Logically it makes sense for it to be `parser`, but I can imagine that you might want to use `hasPragma` inside `insertPragma`, and that would add a dependency on the `parser`.

**Comment by duailibe** [comment]
Well you could write `insertPragma` and `hasPragma` in a module and re-use anything in those, just export accordingly in the `parser`/`printer`, like I did for JavaScript

**Comment by duailibe** [comment]
@azz I think we can merge only the infra and JS enabled and do follow-up PRs for other languages? They aren't working right now anyway...

Sorry for the delay, I was AFK :)

**Comment by azz** [comment]
Sounds good. No need to apologise!

**Comment by duailibe** [comment]
@azz resolved conflicts. I think this is ready for review, two questions:

1) do we need tests for existing `hasPragma` but no `insertPragma` (and vice versa)?
2) should we just continue using jest-docblock for css?

File: index.js [hunk 1]
```diff
@@ -69,7 +62,9 @@ function ensureAllCommentsPrinted(astComments) {
 }
 
 function formatWithCursor(text, opts, addAlignmentSize) {
-  if (opts.requirePragma && !hasPragma(text)) {
+  const selectedParser = parser.resolveParser(opts);
+  const hasPragma = !selectedParser.hasPragma || selectedParser.hasPragma(text);
+  if (opts.requirePragma && !hasPragma) {
     return { formatted: text };
```
Comment (j-f1) [inline index.js]: Should this return the current cursor position as well as the original text?

File: src/language-graphql/index.js [hunk 2]
```diff
@@ -25,6 +27,9 @@ const parsers = {
       return eval("require")("./parser-graphql");
     },
     astFormat: "graphql",
+    hasPragma(text) {
+      return pragmaRe.test(text);
+    },
```
Comment (j-f1) [inline src/language-graphql/index.js]: `hasPragma: text => pragmaRe.test(text),`?

File: src/language-js/index.js [hunk 3]
```diff
@@ -170,18 +172,24 @@ const babylon = {
...
   locEnd
 };
 
 const parsers = {
   babylon,
-  json: babylon,
+  json: Object.assign({}, babylon, {
+    hasPragma() {
+      return false;
+    }
```
Comment (j-f1) [inline src/language-js/index.js]: `hasPragma: null`?

File: tests/insert-pragma/graphql/__snapshots__/jsfmt.spec.js.snap [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+`;
+
+exports[`without-pragma.graphql 1`] = `
+type Foo {
+  one: Type
+}
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+type Foo {
```
Comment (j-f1) [inline tests/insert-pragma/graphql/__snapshots__/jsfmt.spec.js.snap]: Looks like the pragma isn’t getting inserted here.

**Comment by suchipi** [comment]
I'm gonna remove this from the 1.11 release (https://github.com/prettier/prettier/issues/3557#issuecomment-368115155)

File: index.js (hunk 1 above)
Comment (duailibe) [reply index.js]: I don't know, unrelated to this PR

File: tests/insert-pragma/graphql/__snapshots__/jsfmt.spec.js.snap (hunk 4 above)
Comment (duailibe) [reply tests/insert-pragma/graphql/__snapshots__/jsfmt.spec.js.snap]: Yeah I didn't write `insertPragma` for GraphQL yet, I asked for help with this in OP

**Comment by duailibe** [comment]
I removed the GraphQL pragmas since it was incomplete, let's do those in follow-ups.. This includes the infrastructure + JS/JSON pragmas

**Comment by duailibe** [comment]
Does anyone want to review this?

**Comment by duailibe** [comment]
Do we have docs on the plugin interface already? I should probably update those

**Comment by ikatyang** [comment]
--> https://prettier.io/docs/en/plugins.html#developing-plugins

**Pull request merged**